src/lib/sync_handler/sync_handler.ml

```diff
@@ -85,9 +96,117 @@ module Make (Inputs : Inputs_intf) :
       Staged_ledger.pending_coinbase_collection staged_ledger
     in
     (scan_state, merkle_root, pending_coinbases)
+
+  let get_transition_chain ~frontier hashes =
+    Option.all
+    @@ List.map hashes ~f:(fun hash ->
+           Option.merge
+             (Transition_frontier.find frontier hash)
+             (Transition_frontier.find_in_root_history frontier hash)
+             ~f:Fn.const
+           |> Option.map ~f:(fun breadcrumb ->
+                  Transition_frontier.Breadcrumb.transition_with_hash
+                    breadcrumb
+                  |> With_hash.data
+                  |> External_transition.Validated.forget_validation ) )
+
+  module Root = struct
+    let prove ~logger ~frontier seen_consensus_state =
+      let open Option.Let_syntax in
+      let%bind best_tip_with_witness =
+        Best_tip_prover.prove ~logger frontier
+      in
+      let is_tip_better =
+        Consensus.Hooks.select
+          ~logger:
+            (Logger.extend logger [("selection_context", `String "Root.prove")])
+          ~existing:
+            (External_transition.consensus_state best_tip_with_witness.data)
+          ~candidate:seen_consensus_state
+        = `Keep
+      in
+      let%map () = Option.some_if is_tip_better () in
+      best_tip_with_witness
+
+    let verify ~logger ~verifier observed_state peer_root =
+      let open Deferred.Result.Let_syntax in
+      let%bind ( { Coda_intf.Best_tip_verification_result.best_tip=
+                     best_tip, _validation
+                 ; _ } as verification_result ) =
+        Best_tip_prover.verify ~verifier peer_root
+      in
+      let is_before_best_tip candidate =
+        Consensus.Hooks.select
+          ~logger:
+            (Logger.extend logger [("selection_context", `String "Root.verify")])
+          ~existing:(External_transition.consensus_state best_tip.data)
+          ~candidate
+        = `Keep
+      in
+      let%map () =
+        Deferred.return
+          (Result.ok_if_true
+             (is_before_best_tip observed_state)
+             ~error:
+               (Error.createf
+                  !"Peer lied about it's best tip %{sexp:State_hash.t}"
+                  best_tip.hash))
+      in
+      verification_result
+  end
+
+  module Bootstrappable_best_tip = struct
```

Comment:
Can you add doc-comments explaining the `Root` and `Bootstrappable_best_tip` modules? I also think these are not great names because only the interface implies that they are best tip provers, but the names do not describe that.